python/lsst/daf/butler/registry/queries/_results.py

```diff
@@ -304,17 +304,78 @@ def findDatasets(self, datasetType: Union[DatasetType, str], collections: Any, *
         query = builder.finish(joinMissing=False)
         return ParentDatasetQueryResults(db=self._db, query=query, components=components,
-                                         records=self._records)
+                                         records=self._records, datasetType=datasetType)
+
+    def count(self, *, exact: bool = True) -> int:
+        """Count the number of rows this query would return.
+
+        Parameters
+        ----------
+        exact : `bool`, optional
+            If `True`, run the full query and perform post-query filtering if
+            needed to account for that filtering in the count.  If `False`, the
+            result may be an upper bound.
+
+        Returns
+        -------
+        count : `int`
+            The number of rows the query would return, or an upper bound if
+            ``postprocess=False``.
+        """
+        return self._query.count(self._db, exact=exact)
+
+    def any(
+        self, *,
+        execute: bool = True,
+        exact: bool = True,
+    ) -> bool:
+        """Test whether this query returns any results.
+
+        Parameters
+        ----------
+        execute : `bool`, optional
+            If `True`, execute at least a ``LIMIT 1`` query if it cannot be
+            determined prior to execution that the query would return no rows.
+        exact : `bool`, optional
+            If `True`, run the full query and perform post-query filtering if
+            needed, until at least one result row is found.  If `False`, the
+            returned result does not account for post-query filtering, and
+            hence may be `True` even when all result rows would be filtered
+            out.
+
+        Returns
+        -------
+        any : `bool`
+            `True` if the query would (or might, depending on arguments) yield
+            result rows.  `False` if it definitely would not.
+        """
+        return self._query.any(self._db, execute=execute, exact=exact)
+
+    def explain_no_results(self) -> Iterator[str]:
+        """Return human-readable messages that may help explain why the query
+        yields no results.
+
+        Returns
+        -------
+        messages : `Iterator` [ `str` ]
+            String messages that describe reasons the query might not yield any
+            results.
+
+        Notes
+        -----
+        Messages related to post-query filtering are only available if `rows`,
+        `any`, or `count` was already called with the same region (with
+        ``postprocess=True`` for the latter two).
+
+        At present, this method only returns messages that are generated while
+        the query is being built or filtered.  In the future, it may perform
+        its own new follow-up queries, which users may wish to short-circuit
+        simply by not continuing to iterate over its results.
+        """
+        return self._query.explain_no_results(self._db)
```

Comment:
`yield from` instead of `return` (but probably does not matter in this case)?

Reply:
I'm pretty sure that if it does matter, it's only because of performance, and `return` is more likely to win on that question (but this gets back to whether to the iterator vs. generator type covariance question).